osu.Game.Rulesets.Osu/Edit/OsuSelectionHandler.cs

```diff
@@ -207,28 +188,93 @@ public override bool HandleRotation(float delta)
             return true;
         }
 
+        private bool scaleSlider(Slider slider, Vector2 scale)
+        {
+            Quad sliderQuad = getSurroundingQuad(slider.Path.ControlPoints.Select(p => p.Position.Value));
+            Vector2 pathRelativeDeltaScale = new Vector2(1 + scale.X / sliderQuad.Width, 1 + scale.Y / sliderQuad.Height);
+
+            Quad selectionQuad = getSurroundingQuad(new OsuHitObject[] { slider });
+            Quad scaledQuad = new Quad(selectionQuad.TopLeft.X, selectionQuad.TopLeft.Y, selectionQuad.Width + scale.X, selectionQuad.Height + scale.Y);
+            (bool X, bool Y) inBounds = isQuadInBounds(scaledQuad);
+
+            if (!inBounds.X)
+                pathRelativeDeltaScale.X = 1;
+
+            if (!inBounds.Y)
+                pathRelativeDeltaScale.Y = 1;
+
+            foreach (var point in slider.Path.ControlPoints)
+                point.Position.Value *= pathRelativeDeltaScale;
+
+            return true;
+        }
+
+        private bool scaleHitObjects(OsuHitObject[] hitObjects, Anchor reference, Vector2 scale)
+        {
+            // move the selection before scaling if dragging from top or left anchors.
+            if ((reference & Anchor.x0) > 0 && !moveSelection(new Vector2(-scale.X, 0))) return false;
+            if ((reference & Anchor.y0) > 0 && !moveSelection(new Vector2(0, -scale.Y))) return false;
+
+            Quad selectionQuad = getSurroundingQuad(hitObjects);
+            Quad scaledQuad = new Quad(selectionQuad.TopLeft.X, selectionQuad.TopLeft.Y, selectionQuad.Width + scale.X, selectionQuad.Height + scale.Y);
+            (bool X, bool Y) inBounds = isQuadInBounds(scaledQuad);
+
+            foreach (var h in hitObjects)
+            {
+                var newPosition = h.Position;
+
+                // guard against no-ops and NaN.
+                if (scale.X != 0 && selectionQuad.Width > 0 && inBounds.X)
+                    newPosition.X = selectionQuad.TopLeft.X + (h.X - selectionQuad.TopLeft.X) / selectionQuad.Width * (selectionQuad.Width + scale.X);
+
+                if (scale.Y != 0 && selectionQuad.Height > 0 && inBounds.Y)
+                    newPosition.Y = selectionQuad.TopLeft.Y + (h.Y - selectionQuad.TopLeft.Y) / selectionQuad.Height * (selectionQuad.Height + scale.Y);
+
+                h.Position = newPosition;
+            }
+
+            return true;
+        }
+
+        private (bool X, bool Y) isQuadInBounds(Quad quad)
+        {
+            (bool X, bool Y) result;
+
+            result.X = (quad.TopLeft.X >= 0) && (quad.BottomRight.X < DrawWidth);
+            result.Y = (quad.TopLeft.Y >= 0) && (quad.BottomRight.Y < DrawHeight);
+
+            return result;
+        }
+
         private bool moveSelection(Vector2 delta)
         {
             var hitObjects = selectedMovableObjects;
 
+            foreach (var h in hitObjects)
+                h.Position += delta;
+
+            return true;
+        }
```

Comment:
Should not return a bool if it'll always return `true`. In fact I'm not sure whether this should exist at all anymore (all movement should be most likely always happening in bounds).